### ISSUE TITLE
OIDC prompt login

### ISSUE THREAD

**Issue by james-d-elliott** [authelia/authelia#2596]
### Description

OpenID Connect workflow doesn't seem to do the appropriate thing when performing propmt type of login. ~~This will take investigation to narrow down what is going wrong and what we need to do. Example that may be useful is Portainer.~~

The reason this occurs is because the auth_time value needs to be updated, this can be done via another authentication which complies with the spec.  Probably the best way to achieve this is to detect the request for a login prompt and provide the one relevant to the client i.e. if the client is configured as 1FA provide a password prompt, if the client is configured for 2FA allow a 2FA auth.

### Expected Behaviour

_N/A_

### Reproduction Steps

_N/A_

### Additional Information

https://github.com/authelia/authelia/issues/189#issuecomment-966181727

Basically it's occurring on this line in fosite: https://github.com/ory/fosite/blob/v0.40.2/handler/openid/validator.go#L148-L152

auth_time claim < rat claim

**Comment by rishabhkalra** [comment]
I'm actually seeing this same behavior with proxmox. Proxmox is making a request with `prompt=login` but I only see a consent screen and I'm not re-authenticated when I accept and get redirected back to proxmox, seeing the same error described in the https://github.com/authelia/authelia/issues/189#issuecomment-966181727 .

**Comment by james-d-elliott** [comment]
I did some research on this, basically it appears that we should perform authentication again for the user. However this is not currently something we support.. I think this probably is going to require a fairly large refactor of the main workflow. 

See here for spec:

https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest

Specifically:

```
prompt
    OPTIONAL. Space delimited, case sensitive list of ASCII string values that specifies whether the Authorization Server prompts the End-User for reauthentication and consent. The defined values are:

        none
            The Authorization Server MUST NOT display any authentication or consent user interface pages. An error is returned if an End-User is not already authenticated or the Client does not have pre-configured consent for the requested Claims or does not fulfill other conditions for processing the request. The error code will typically be login_required, interaction_required, or another code defined in Section 3.1.2.6. This can be used as a method to check for existing authentication and/or consent. 
        login
            The Authorization Server SHOULD prompt the End-User for reauthentication. If it cannot reauthenticate the End-User, it MUST return an error, typically login_required. 
        consent
            The Authorization Server SHOULD prompt the End-User for consent before returning information to the Client. If it cannot obtain consent, it MUST return an error, typically consent_required. 
        select_account
            The Authorization Server SHOULD prompt the End-User to select a user account. This enables an End-User who has multiple accounts at the Authorization Server to select amongst the multiple accounts that they might have current sessions for. If it cannot obtain an account selection choice made by the End-User, it MUST return an error, typically account_selection_required. 

    The prompt parameter can be used by the Client to make sure that the End-User is still present for the current session or to bring attention to the request. If this parameter contains none with any other value, an error is returned. 
```

**Comment by james-d-elliott** [comment]
As a small aside we also need to cater for all those other prompt types.

**Comment by peterbuga** [comment]
@YouveGotMeowxy i cannot comment on original thread https://github.com/authelia/authelia/issues/3396 due to lock but i've come across the same problem:
in my case the part i overlooked is that I was having the domain `portainer.example.com` to be already behind authelia, so 1st i would login to authelia => get to portainer login page, select OAuth login and get the error.
the fix was to disable authelia for portainer subdomain and login would work as expected, but i didn't want the portainer login to be exposed to internet so flipping the switch in portainer settings would get me straight to authelia login page
```
Single Sign-On

Use SSO [x]
Hide internal authentication prompt [x]```

hope this helps anyone

**Comment by blysik** [comment]
Hi, I am a new Authelia user, and believe I've setup OpenID with Portainer correctly, per the documentation.  I was initially getting the timing error which complained about time, but now I receive the following error:

```
time="2023-03-11T01:02:04Z" level=error msg="Access Request failed with error: Client authentication failed (e.g., unknown client, no client authentication included, or unsupported authentication method). the passwords don't match" method=POST path=/api/oidc/token remote_ip=172.18.0.1 stack="github.com/authelia/authelia/v4/internal/handlers/handler_oidc_token.go:27 
```

Should I just shelve Portainer auth via OpenId for now and wait?  Or should this work and I'm just doing something wrong?  Thanks!

**Comment by james-d-elliott** [comment]
This topic is for discussion about the prompt login feature, I'm locking it as it keeps going off topic. The above error indicates the client application sent the wrong client secret (password). Open a discussion if you're still having issues.
